### ISSUE TITLE
How to get dotnet array from torch tensor

### ISSUE THREAD

**Issue by ChengYen-Tang** [dotnet/TorchSharp#863]
Hello everyone,
How can I get dotnet array from torch tensor

**Comment by NiklasGustafsson** [comment]
Try: tensor.data()<T>.ToArray()

**Comment by ChengYen-Tang** [comment]
1. What type should this Tensor be? `nn.utils.parameters_to_vector(parameters()).detach().cpu()`
  - float[] parameters = nn.utils.parameters_to_vector(parameters()).detach().cpu().data<float>().ToArray();
  - double[] parameters = nn.utils.parameters_to_vector(parameters()).detach().cpu().data<double>().ToArray();
1. What to do if it is an N-dimensional array.

**Comment by NiklasGustafsson** [comment]
1. in tensor.data<T>().ToArray(), T should be the .NET type that corresponds to the tensor's element type.

2. You can only get a 1-dimensional array out.

**Comment by ChengYen-Tang** [comment]
1. Is nn.Module.parameters() a 1D array? and what is its default type?
2. Maybe TorchSharp can add a method ToSystemArray(), for developers to perform security transformation.

like this.
```
Array input = new float[3,4,5];
Tensor tensor = from_array(input);
float[,,] result = tensor.data<float>().ToSystemArray() as float[,,];
```

**Comment by NiklasGustafsson** [comment]
>> Is nn.Module.parameters() a 1D array?

It's an `IEnumerable<Parameter>`, where `Parameter` is a subclass of `Tensor`, so it can be (and usually is) N-dim. Each parameter in the enumeration has independent dimensions.

**Comment by ChengYen-Tang** [comment]
> 2. Maybe TorchSharp can add a method ToSystemArray(), for developers to perform security transformation.
So is there any chance of this requirement being fulfilled?
If I can, I don't seem to have to worry about the following problems.
> 1. Is nn.Module.parameters() a 1D array? and what is its default type?

**Comment by NiklasGustafsson** [comment]
@ChengYen-Tang -- I don't know exactly what you are asking for. 

What does ToSystemArray() do differently from ToArray<T>()? What type does it return?

Can you elaborate on your need to get a N-dimensional .NET array out of a tensor? What is the scenario? TorchSharp is intended to be used in a way where you get your numerical data into tensors and then keep working with that as your fundamental data structure.

**Comment by NiklasGustafsson** [comment]
I see. 

System.Array is an abstract type, so we would still have to determine a concrete type to return.

We could have a ToNDArray() which would statically return 'System.Array' but dynamically an N-dimensional array with the same N as the tensor. I can't guarantee that it would be fast, but it's definitely doable.

I'm going to take some vacation, so I won't get to this until January. If you want to contribute a solution, it would be going in TensorAccessor.cs, and look something like this:

```C#
        public System.Array ToNDArray()
        {
            System.Array result = null;
            var shape = _tensor.shape;
            switch (_tensor.ndim) {
            default:
                return ToArray();
            case 2:
                result = new T[shape[0], shape[1]];
                // Fill data from _tensor
                return result;
            case 3:
                result = new T[shape[0], shape[1], shape[2]];
                // Fill data from _tensor
                return result;
            case 4:
                result = new T[shape[0], shape[1], shape[2], shape[3]];
                // Fill data from _tensor
                return result;
            case 5:
                result = new T[shape[0], shape[1], shape[2], shape[3], shape[4]];
                // Fill data from _tensor
                return result;
            case 6:
                result = new T[shape[0], shape[1], shape[2], shape[3], shape[4], shape[5]];
                // Fill data from _tensor
                return result;
            }
        }
```

The trick to getting it to perform well will be in how the 'Fill data from _tensor' is done. There's a lot of possibilities for wasted cycles and memory copies there.

**Comment by NiklasGustafsson** [comment]
Something like this should work, but it'll be awfully slow and generate lots of temps:

```C#
default:
                throw new NotImplementedException("ToNDArray() for more than '6' dimensions.");
            case 0:
            case 1:
                return ToArray();
            case 2: {
                    var result = new T[shape[0], shape[1]];
                    for (int i0 = 0; i0 < shape[0]; i0++) {
                        for (int i1 = 0; i1 < shape[1]; i1++) {
                            result[i0, i1] = _tensor[i0,i1].item<T>();
                        }
                    }
                    return result;
                }
```

**Comment by NiklasGustafsson** [comment]
@ChengYen-Tang, @xhuan8 -- maybe you can take a look at PR #868 and give me feedback?

**Comment by ChengYen-Tang** [comment]
@NiklasGustafsson, I opened a pr to your branch. https://github.com/NiklasGustafsson/TorchSharp/pull/9

**Comment by NiklasGustafsson** [comment]
@ChengYen-Tang --  I saw your PR to my personal repo. Thanks for generalizing. @xhuan8 is right in his comment on the PR, TranslateIndex is far too expensive to call for every element. I'll rewrite my code (probably tomorrow), and then maybe you can do the same generalization again?
